t/t7519-status-fsmonitor.sh

```diff
@@ -354,4 +354,18 @@ test_expect_success 'discard_index() also discards fsmonitor info' '
 	test_cmp expect actual
 '
 
+# Use test files that start with 'z' so that the entries being added
+# and removed appear at the end of the index.
+test_expect_success 'compare status before and after staging/unstaging ' '
+	test_commit initial &&
+	removed=$(seq 1 100 | sed "s/^/z/") &&
+	touch $removed &&
+	git status >expected &&
+	git add $removed &&
+	test_config core.fsmonitor "$TEST_DIRECTORY/t7519/fsmonitor-env" &&
+	FSMONITOR_LIST="$removed" git restore -S $removed &&
+	FSMONITOR_LIST="$removed" git status >actual &&
+	test_cmp expected actual
```

Comment:
@dscho I made a few tweaks here, please let me know what you think.  

I changed the description and had the test validate `status` because I wasn't sure the best way to demonstrate that the test was actually validating that fsmonitor was ignoring the removed entries.

I confirmed the test fails without the fix.

Reply:
Yes, you introduced the `expected` vs `actual` test. I like it! To keep with the convention, could you rename it to `expect`? That's how it is called in most places in Git's test suite.

Reply:
@dscho I hit a failure on `osx-clang` after pushing my latest set of changes (the other platforms were fine), here is the diff of `git-status`:

```
2019-09-30T17:47:34.6313770Z ++ diff -u expect actual
2019-09-30T17:47:34.6314380Z --- expect	2019-09-30 17:42:28.000000000 +0000
2019-09-30T17:47:34.6315030Z +++ actual	2019-09-30 17:42:28.000000000 +0000
2019-09-30T17:47:34.6316420Z @@ -11,10 +11,6 @@
2019-09-30T17:47:34.6316690Z    (use "git add <file>..." to include in what will be committed)
2019-09-30T17:47:34.6316900Z  	after
2019-09-30T17:47:34.6317070Z  	before
2019-09-30T17:47:34.6317580Z -	dir1/new
2019-09-30T17:47:34.6318160Z -	dir1/untracked
2019-09-30T17:47:34.6318720Z -	dir2/new
2019-09-30T17:47:34.6319280Z -	dir2/untracked
2019-09-30T17:47:34.6319820Z  	dot-git/
2019-09-30T17:47:34.6320020Z  	new
2019-09-30T17:47:34.6320540Z  	trace-after
2019-09-30T17:47:34.6320760Z error: last command exited with $?=1
```
https://gvfs.visualstudio.com/23a36a33-54f4-42be-84ae-6286f1f5fd6a/_apis/build/builds/16835/logs/160

Any guesses as to what might have caused it? (the `dir1` and `dir2` files are created by a different test).

This is a bit strange because:
- The tests passed fine last week: https://gvfs.visualstudio.com/ci/_build/results?buildId=16778&view=results
- All of the test files that this test uses appear well after the `dir` files in the index
- The test passed when I re-ran it: https://gvfs.visualstudio.com/ci/_build/results?buildId=16835&view=logs&j=bb92627a-0f17-5507-454a-87285e0fd6b8 and I'm having trouble seeing what would make this test racy.  

I did see that the time timing of when t7519 ran was different across runs, and I'm not sure if that could be related.

Do you have any advice on how to proceed?  I have been unable to repro this failure locally and thus far have only seen it once on the build machines.

cc: @derrickstolee @jeffhostetler for their thoughts as well

Reply:
I'm having a little trouble following what the test is doing, so I'm going to be vague here. :-)

On line 363, we have "git status >expect" and on line 367 "FSMONITOR... git status >actual".
This threw me a minute -- I'm was thinking that "expect" would be a test constant. 

Perhaps it would be better to explicitly (hard-code) the correct answer using a here-doc (the
"cat... EOF" thing).  And then have these 2 status commands be "actual1" and "actual2" and
compare each of them against the well-known "expect".   Then at least we'd know which side
was wrong.

There may be a race in here, but I don't know enough about the fsmonitor code to comment
right now.

Reply:
> This threw me a minute -- I'm was thinking that "expect" would be a test constant.
>
>Perhaps it would be better to explicitly (hard-code) the correct answer using a here-doc (the "cat... EOF" thing). And then have these 2 status commands be "actual1" and "actual2" and compare each of them against the well-known "expect". Then at least we'd know which side was wrong.

I was trying to use test code as an example when I made the change:

```
# Status is well tested elsewhere so we'll just ensure that the results are
# the same when using core.fsmonitor.
test_expect_success 'compare status with and without fsmonitor' '
	write_integration_script &&
	clean_repo &&
	dirty_repo &&
	git add new &&
	git add dir1/new &&
	git add dir2/new &&
	git status >actual &&
	git -c core.fsmonitor= status >expect &&
	test_i18ncmp expect actual
```

Although I now see that if we do end up sticking with the current approach that `test_i18ncmp` should be used instead of `test_cmp`.  

The messy part I've found with trying to declare an "expected" constant is the expected output will change base on which of the tests cases have run prior to running this test case (and would it be okay for the expected test case to always assume all previous tests have run)?

I'm wondering if I should revert it to the original proposed by @dscho that did not include a comparison of the output of status.  After trying a bunch of iterations I can confirm that to repro the segfault without the fix in this PR the `test_config core.fsmonitor` line must be set after running the initial `git status` and `git add`, but I have not been able to figure out why we must wait until after those git commands have run.  

In other words, I'm not sure why I can't hit a repro when when I set the `core.fsmonitor` ahead of the initial `git status` and `git add`.  @dscho how did you track down this ordering down?

Related to the earlier discussion regarding expected output, in general I'm having some trouble figuring out how each of these test cases impacts the next.  For example:

- `./t7519-status-fsmonitor.sh --run=29` fails without the fix
- `./t7519-status-fsmonitor.sh --run=1,29` passes without the fix
- `./t7519-status-fsmonitor.sh --run=1,28,29` fails without the fix

Is there any general guidance as to whether all test cases should be able to run individually?  Or should developers assume that all test cases will be run together?

Reply:
> I'm wondering if I should revert it to the original proposed by @dscho that did not include a comparison of the output of status. 

I pushed a second commit with that approach here:

https://github.com/microsoft/git/pull/201/commits/11ff0f5ef30bb635551581af4c27964a4cec5a2e

Reply:
I'm not sure about the larger questions you're asking here, but you might look at "test_when_finished" clause that can cleanup per-test cruft and eliminate some of
these dependencies.